### PR TITLE
[SYCL] Get rid of using deprecated API in SYCL RT and tests

### PR DESCRIPTION
**Pull request by psamolysov-intel**
The sycl::backend_return_t type alias is used as return type for all
the (deprecated) SyclObj.get_native() functions instead of the
sycl::detail::interop one. Also all the invokes of the SyclObj.get_native()
methods in the unit tests are replaced with the sycl::get_native()
free function.

All the unit tests use the new INSTANTIATE_TEST_SUITE_P googletest
macro instead of the deprecated INSTANTIATE_TEST_CASE_P one.

Usages of the deprecated function interop_task  are replaced with
host_task in unit tests.

### CODE REVIEW

**Review by vladimirlaz — DISMISSED** [review]

**Review by vladimirlaz — APPROVED** [review]

File: sycl/unittests/assert/assert.cpp [hunk 1]
```diff
@@ -357,12 +359,14 @@ static pi_result redefinedProgramGetInfo(pi_program P,
   if (PI_PROGRAM_INFO_DEVICES == ParamName) {
     EXPECT_EQ(ParamValueSize, 1 * sizeof(cl_device_id));
 
-    cl_device_id Dev = Device->get_native<sycl::backend::opencl>();
+    cl_device_id Dev = sycl::get_native<sycl::backend::opencl>(*Device);
 
-    if (ParamValue)
+    if (ParamValue) {
```
Comment (steffenlarsen) [inline sycl/unittests/assert/assert.cpp:364]: According to the [LLVM Coding Standard](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements) braceless is preferred for single-expression control-flow statements, unless it makes it harder to read. Personally I don't mind and I don't think we adhere to it many places, but it is something to think about.

**Review by steffenlarsen — APPROVED** [review]
LGTM!

Maybe it would make sense to change the names of the tests that use kernel-bundle instead of program, but it is minor.

File: sycl/unittests/assert/assert.cpp (hunk 1 above)
Comment (psamolysov-intel) [reply sycl/unittests/assert/assert.cpp:364]: @steffenlarsen Thank you for the comment, I won't insert `{ }` in one-line control-flow statements anymore.

File: sycl/include/CL/sycl/context.hpp [hunk 2]
```diff
@@ -217,12 +217,10 @@ class __SYCL_EXPORT context {
...
   ///
   /// \return a native handle, the type of which defined by the backend.
-  template <backend BackendName>
+  template <backend Backend>
   __SYCL_DEPRECATED("Use SYCL 2020 sycl::get_native free function")
-  auto get_native() const ->
-      typename detail::interop<BackendName, context>::type {
-    return reinterpret_cast<
-        typename detail::interop<BackendName, context>::type>(getNative());
+  auto get_native() const -> backend_return_t<Backend, context> {
```
Comment (bader) [inline sycl/include/CL/sycl/context.hpp:222]:
```suggestion
  backend_return_t<Backend, context> get_native() const {
```
I found trailing return type less readable. There is no reason to use it in this case and SYCL spec also doesn't use it.

File: sycl/include/CL/sycl/kernel.hpp [hunk 3]
```diff
@@ -202,11 +202,8 @@ class __SYCL_EXPORT kernel {
 
   template <backend Backend>
   __SYCL_DEPRECATED("Use SYCL 2020 sycl::get_native free function")
-  typename backend_traits<Backend>::template return_type<kernel> get_native()
-      const {
-    return detail::pi::cast<
-        typename backend_traits<Backend>::template return_type<kernel>>(
-        getNativeImpl());
+  auto get_native() const -> backend_return_t<Backend, kernel> {
+    return detail::pi::cast<backend_return_t<Backend, kernel>>(getNativeImpl());
```
Comment (bader) [inline sycl/include/CL/sycl/kernel.hpp:206]: Just out of curiosity: why kernel uses `getNativeImpl`, when all other classes use `getNative`?

File: sycl/unittests/assert/assert.cpp (hunk 1 above)
Comment (bader) [reply sycl/unittests/assert/assert.cpp:364]: @psamolysov-intel, please, revert this change.

File: sycl/include/CL/sycl/context.hpp (hunk 2 above)
Comment (psamolysov-intel) [reply sycl/include/CL/sycl/context.hpp]: @bader Thank you for the suggestion. I've applied the change for this header as well as the others that define the `get_native()` method.

File: sycl/unittests/assert/assert.cpp (hunk 1 above)
Comment (psamolysov-intel) [reply sycl/unittests/assert/assert.cpp]: Reverted.

File: sycl/include/CL/sycl/kernel.hpp (hunk 3 above)
Comment (psamolysov-intel) [reply sycl/include/CL/sycl/kernel.hpp]: The `getNativeImpl` method is renamed into the `getNative` as the method is named in all other classes.

**Review by vladimirlaz — APPROVED** [review]

**Review by steffenlarsen — DISMISSED** [review]

**Comment by bader** [comment]
```
/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.src/sycl/source/kernel.cpp:131:18: error: no declaration matches 'pi_native_handle cl::sycl::kernel::getNativeImpl() const'
  131 | pi_native_handle kernel::getNativeImpl() const { return impl->getNative(); }
      |                  ^~~~~~
```
@psamolysov-intel, please, fix build errors.

**Comment by psamolysov-intel** [comment]
@bader Fixed, thanks.

**Comment by psamolysov-intel** [comment]
@bader ABI Test failed, since the symbol `_ZNK2cl4sycl6kernel13getNativeImplEv` is missing. Can we just change the symbol name in the test since the `getNativeImpl` method is private in the `kernel` class?

**Comment by bader** [comment]
> @bader ABI Test has failed since the symbol `_ZNK2cl4sycl6kernel13getNativeImplEv` is missing. Can we just change the symbol name in the test since the `getNativeImpl` method is private in the `kernel` class?

I don't think so. This seems to be an ABI breaking change. I suppose application built with older DPC++ versions won't work with the runtime after such change.
@romanovvlad, @alexbatashev, can we add `getNative` member in addition to `getNativeImpl` and deprecate `getNativeImpl` with intention to remove it later?

**Comment by alexbatashev** [comment]
> > @bader ABI Test has failed since the symbol `_ZNK2cl4sycl6kernel13getNativeImplEv` is missing. Can we just change the symbol name in the test since the `getNativeImpl` method is private in the `kernel` class?
> 
> 
> 
> I don't think so. This seems to be an ABI breaking change. I suppose application built with older DPC++ versions won't work with the runtime after such change.
> 
> @romanovvlad, @alexbatashev, can we add `getNative` member in addition to `getNativeImpl` and deprecate `getNativeImpl` with intention to remove it later?

That would make sense

**Comment by psamolysov-intel** [comment]
@bader @alexbatashev I've reverted the `getNativeImpl()` method and marked it as deprecated.

**Review by bader — APPROVED** [review]

**Review by steffenlarsen — APPROVED** [review]

**Pull request merged**